lib/interferon.rb

```diff
@@ -26,15 +26,20 @@ def initialize(alerts_repo_path, groups_sources, host_sources, destinations)
       @groups_sources = groups_sources
       @host_sources = host_sources
       @destinations = destinations
+      @request_shutdown = false
     end
 
     def run(dry_run = false)
+      Signal.trap("TERM") do
+        log.info "SIGTERM received. shutting down gracefully..."
+        @request_shutdown = true
+      end
       run_desc = dry_run ? 'dry run' : 'run'
       log.info "beginning alerts #{run_desc}"
 
       alerts = read_alerts
       groups = read_groups(@groups_sources)
-      hosts = read_hosts(@host_sources)
+      hosts = read_hosts(@host_sources) unless @request_shutdown
```

Comment:
i don't think you should have this both here, and inside the main loop of that function. omit the call here, since you cannot avoid it inside the function.